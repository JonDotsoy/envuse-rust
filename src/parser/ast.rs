use std::collections::BTreeMap;

use serde::{Serialize, Deserialize};

use crate::syntax_error::SyntaxError;

use super::{span::Span, tokenizer::Token};

/// Expression
///
/// Used
#[derive(Debug, Serialize, Deserialize)]
pub enum Expression {
    Document {
        span: Span,
        executable: Option<String>,
        elements: Vec<Expression>,
    },
    CommentBlock {
        span: Span,
        raw: Vec<String>,
    },
    Variable {
        span: Span,
        comment: Box<Option<Expression>>,
        name: String,
        variable_type: Option<String>,
        options_variable_type: Option<BTreeMap<String, Option<Expression>>>,
        default_value: Box<Option<Expression>>,
        nullable: bool,
    },
    DefaultValue {
        span: Span,
        value: String,
    },
    OptionValue {
        span: Span,
        value: String,
    },
}

impl Expression {
    fn to_span(&self) -> Span {
        match self {
            Expression::Document { span, .. } => span.clone(),
            Expression::CommentBlock { span, .. } => span.clone(),
            Expression::Variable { span, .. } => span.clone(),
            Expression::DefaultValue { span, .. } => span.clone(),
            Expression::OptionValue { span, .. } => span.clone(),
        }
    }
}

#[derive(Debug, Clone)]
struct Cursor<T>
where
    T: AsRef<Vec<Token>>,
{
    index: usize,
    items: T,
}

impl<T> Cursor<T>
where
    T: AsRef<Vec<Token>>,
{
    fn new(items: T) -> Self {
        Self { index: 0, items }
    }

    fn assert_current_kind<S: ToString, const U: usize>(
        &self,
        kinds: [S; U],
    ) -> Result<&Token, SyntaxError> {
        let kinds_str = kinds.map(|k| k.to_string());
        if let Some(token) = self.select_current() {
            if kinds_str.contains(&token.kind) {
                Ok(token)
            } else {
                Err(SyntaxError::new(
                    format!("Unexpected token {}", token.kind),
                    token.span.clone(),
                ))
            }
        } else {
            Err(SyntaxError::new(
                "Unexpected token",
                Span {
                    start: self.index,
                    end: self.index + 1,
                },
            ))
        }
    }

    fn forward_some_kind<S: ToString, const Z: usize>(&mut self, kinds: [S; Z]) {
        let kinds_str = kinds.map(|k| k.to_string());
        while let Some(token) = self.select_current() {
            if kinds_str.contains(&token.kind) {
                self.forward(1);
                continue;
            }
            break;
        }
    }

    fn select_current(&self) -> Option<&Token> {
        self.items.as_ref().get(self.index)
    }

    fn select_next(&self) -> Option<&Token> {
        self.items.as_ref().get(self.index + 1)
    }

    fn select_nexts(&self, steps: usize) -> Option<&Token> {
        self.items.as_ref().get(self.index + steps)
    }

    fn select_prev(&self) -> Option<&Token> {
        self.items.as_ref().get(self.index - 1)
    }

    fn next(&mut self) -> Option<&Token> {
        self.forward(1);
        self.select_prev()
    }

    fn forward(&mut self, steps: usize) {
        self.index = self.index + steps;
    }
}

pub struct AST {}

impl AST {
    pub fn parse(tokens: Vec<Token>) -> Result<Expression, SyntaxError> {
        let mut executable: Option<String> = None;
        let ref mut token_cursor = Cursor::new(tokens);
        let mut span_start: usize = 0;
        let mut span_end: usize = 0;

        if let Some(token) = token_cursor.select_current() {
            if token.kind == "comment" && token.raw.starts_with("#!") {
                span_start = token.span.start;
                span_end = token.span.end;
                executable = Some(token.raw.to_string());
                token_cursor.forward(1);
            }
        }

        let elements = Self::parse_expressions_list(token_cursor)?;

        if let Some(expression) = elements.last() {
            span_end = expression.to_span().end;
        }

        Ok(Expression::Document {
            executable,
            elements,
            span: Span {
                start: span_start,
                end: span_end,
            },
        })
    }

    fn parse_expressions_list(
        tokens_cursor: &mut Cursor<Vec<Token>>,
    ) -> Result<Vec<Expression>, SyntaxError> {
        let mut vec: Vec<Expression> = vec![];

        while let Some(token) = tokens_cursor.select_current() {
            if token.kind == "newline" || token.kind == "space" {
                tokens_cursor.forward(1);
                continue;
            }
            if token.kind == "comment" {
                vec.push(Self::parse_block_comment(tokens_cursor)?);
                continue;
            }
            if token.kind == "keyword" {
                vec.push(Self::parse_variable(tokens_cursor, None)?);
                continue;
            }
            return Err(SyntaxError::new("Unexpected type", token.span.clone()));
        }

        Ok(vec)
    }

    fn parse_block_comment(
        tokens_cursor: &mut Cursor<Vec<Token>>,
    ) -> Result<Expression, SyntaxError> {
        let span_start = tokens_cursor.select_current().unwrap().span.start;
        let mut span_end = tokens_cursor.select_current().unwrap().span.end;
        let mut raw: Vec<String> = vec![];

        while let Some(token) = tokens_cursor.select_current() {
            if token.kind == "space" {
                tokens_cursor.forward(1);
                continue;
            }

            if token.kind != "comment" {
                break;
            }

            raw.push(token.raw.to_string());
            span_end = token.span.end;

            if let Some(token) = tokens_cursor.select_next() {
                if token.kind == "newline" {
                    tokens_cursor.forward(1);
                }
            }

            tokens_cursor.forward(1);
        }

        let comment = Expression::CommentBlock {
            span: Span {
                start: span_start,
                end: span_end,
            },
            raw,
        };

        if let Some(token) = tokens_cursor.select_current() {
            if token.kind == "keyword" {
                return Ok(Self::parse_variable(tokens_cursor, Some(comment))?);
            }
        }

        Ok(comment)
    }

    fn parse_variable(
        tokens_cursor: &mut Cursor<Vec<Token>>,
        comment: Option<Expression>,
    ) -> Result<Expression, SyntaxError> {
        // Assert Start with keyword
        tokens_cursor.assert_current_kind(["keyword"])?;

        let name: String = tokens_cursor.select_current().unwrap().raw.to_string();
        let span_start: usize = tokens_cursor.select_current().unwrap().span.start;
        let mut span_end: usize = tokens_cursor.select_current().unwrap().span.end;
        let mut variable_type: Option<String> = None;
        let mut options_variable_type: Option<BTreeMap<String, Option<Expression>>> = None;
        let mut default_value: Option<Expression> = None;
        let mut nullable: bool = false;
        tokens_cursor.forward(1);

        tokens_cursor.forward_some_kind(["space"]);

        if tokens_cursor.assert_current_kind(["colon"]).is_ok() {
            tokens_cursor.forward(1);
            tokens_cursor.forward_some_kind(["space"]);
            tokens_cursor.assert_current_kind(["keyword"])?;
            let variable_type_token = tokens_cursor.select_current().unwrap();
            variable_type = Some(variable_type_token.raw.to_string());
            span_end = variable_type_token.span.end;
            tokens_cursor.forward(1);
            tokens_cursor.forward_some_kind(["space"]);

            if tokens_cursor.assert_current_kind(["less_than"]).is_ok() {
                tokens_cursor.forward(1);
                tokens_cursor.forward_some_kind(["space", "newline"]);

                let mut options: BTreeMap<String, Option<Expression>> = Default::default();

                while tokens_cursor.assert_current_kind(["keyword"]).is_ok() {
                    let option_key = tokens_cursor.select_current().unwrap().raw.to_string();
                    tokens_cursor.forward(1);
                    tokens_cursor.forward_some_kind(["space", "newline"]);
                    if tokens_cursor.assert_current_kind(["equal"]).is_ok() {
                        tokens_cursor.assert_current_kind(["equal"])?;
                        tokens_cursor.forward(1);
                        tokens_cursor.forward_some_kind(["space"]);
                        let value_expression = Self::parse_items_option_value(tokens_cursor)?;
                        options.insert(option_key, Some(value_expression));
                        tokens_cursor.forward_some_kind(["space", "newline"]);
                    } else {
                        options.insert(option_key, None);
                    }
                }

                let greater_than = tokens_cursor.assert_current_kind(["greater_than"])?;
                span_end = greater_than.span.end;
                tokens_cursor.forward(1);

                options_variable_type = Some(options);

                tokens_cursor.forward_some_kind(["space"]);
            }

            if let Ok(token) = tokens_cursor.assert_current_kind(["question_mark"]) {
                span_end = token.span.end;
                tokens_cursor.forward(1);
                nullable = true;
            }
        }

        if tokens_cursor.assert_current_kind(["equal"]).is_ok() {
            tokens_cursor.forward(1);
            tokens_cursor.forward_some_kind(["space"]);
            let default_value_expression = Self::parse_items_default_value(tokens_cursor)?;
            span_end = default_value_expression.to_span().end;
            default_value = Some(default_value_expression);
        }

        Ok(Expression::Variable {
            span: Span {
                start: span_start,
                end: span_end,
            },
            comment: Box::new(comment),
            name: name,
            variable_type,
            options_variable_type,
            default_value: Box::new(default_value),
            nullable,
        })
    }

    fn parse_items_default_value(
        tokens_cursor: &mut Cursor<Vec<Token>>,
    ) -> Result<Expression, SyntaxError> {
        match tokens_cursor.select_current() {
            Some(token) if token.kind == "string" => {
                Ok(Self::parse_items_default_value_string(tokens_cursor)?)
            }
            Some(token) if token.kind == "number" => {
                Ok(Self::parse_items_default_value_number(tokens_cursor)?)
            }
            Some(token) => Err(SyntaxError::new("Unexpected token", token.span.clone())),
            _ => {
                Err(SyntaxError::new(
                    "Unexpected token",
                    Span {
                        start: tokens_cursor.index,
                        end: tokens_cursor.index + 1,
                    },
                ))
                // panic!("Unexpected token");
            }
        }
    }

    fn parse_items_default_value_string(
        tokens_cursor: &mut Cursor<Vec<Token>>,
    ) -> Result<Expression, SyntaxError> {
        let token = tokens_cursor.select_current().unwrap();
        let a = Expression::DefaultValue {
            span: token.span.clone(),
            value: token.raw.to_string(),
        };
        tokens_cursor.forward(1);
        return Ok(a);
    }

    fn parse_items_default_value_number(
        tokens_cursor: &mut Cursor<Vec<Token>>,
    ) -> Result<Expression, SyntaxError> {
        let token = tokens_cursor.select_current().unwrap();
        let a = Expression::DefaultValue {
            span: token.span.clone(),
            value: token.raw.to_string(),
        };
        tokens_cursor.forward(1);
        return Ok(a);
    }

    fn parse_items_option_value(
        tokens_cursor: &mut Cursor<Vec<Token>>,
    ) -> Result<Expression, SyntaxError> {
        match tokens_cursor.select_current() {
            Some(token) if token.kind == "string" => {
                Ok(Self::parse_items_option_value_string(tokens_cursor)?)
            }
            Some(token) if token.kind == "number" => {
                Ok(Self::parse_items_option_value_number(tokens_cursor)?)
            }
            Some(token) => Err(SyntaxError::new("Unexpected token", token.span.clone())),
            _ => {
                Err(SyntaxError::new(
                    "Unexpected token",
                    Span {
                        start: tokens_cursor.index,
                        end: tokens_cursor.index + 1,
                    },
                ))
                // panic!("Unexpected token");
            }
        }
    }

    fn parse_items_option_value_string(
        tokens_cursor: &mut Cursor<Vec<Token>>,
    ) -> Result<Expression, SyntaxError> {
        let token = tokens_cursor.select_current().unwrap();
        let a = Expression::OptionValue {
            span: token.span.clone(),
            value: token.raw.to_string(),
        };
        tokens_cursor.forward(1);
        return Ok(a);
    }

    fn parse_items_option_value_number(
        tokens_cursor: &mut Cursor<Vec<Token>>,
    ) -> Result<Expression, SyntaxError> {
        let token = tokens_cursor.select_current().unwrap();
        let a = Expression::OptionValue {
            span: token.span.clone(),
            value: token.raw.to_string(),
        };
        tokens_cursor.forward(1);
        return Ok(a);
    }
}
