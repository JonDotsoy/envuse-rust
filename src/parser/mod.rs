use self::{
    ast::{Expression, AST},
    tokenizer::Tokenizer,
};

pub mod ast;
pub mod tokenizer;

/// Parse source
pub fn parse<A: ToString>(payload: A) -> Result<Expression, ()> {
    let tokens = Tokenizer::parse(payload.to_string());
    Ok(AST::parse(tokens))
}
