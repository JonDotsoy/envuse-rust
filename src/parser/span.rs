use serde::{Serialize, Deserialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl Span {
    pub fn substring<T: ToString>(&self, payload: T) -> String {
        payload.to_string()[self.start..self.end].to_string()
    }
}
