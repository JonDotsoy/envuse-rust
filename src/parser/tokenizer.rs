use std::{ops::Add, usize, vec};

#[derive(Debug)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

#[derive(Debug)]
pub struct Token {
    pub kind: String,
    pub raw: String,
    pub span: Span,
}

impl Token {
    fn new<A: ToString>(kind: A, cursor: &Cursor, span: Span) -> Self {
        Self {
            kind: kind.to_string(),
            raw: cursor.get_by_span(&span),
            span,
        }
    }
}

#[derive(Debug)]
struct Cursor {
    payload: String,
    index: usize,
}

impl Cursor {
    fn new<A>(index: usize, payload: A) -> Self
    where
        A: ToString,
    {
        Self {
            payload: payload.to_string(),
            index,
        }
    }

    fn has_next<'a>(&'a self) -> bool {
        (self.payload.len() > (self.index + 1)).clone()
    }

    fn has_current<'a>(&'a self) -> bool {
        self.payload.len() > self.index
    }

    fn current_matches_char(&self, val: char) -> bool {
        if let Some(c) = self.current_char() {
            c == val
        } else {
            false
        }
    }

    fn next_matches_char(&self, val: char) -> bool {
        if let Some(c) = self.next_char() {
            c == val
        } else {
            false
        }
    }

    fn next_matches_range_char(&self, range: std::ops::RangeInclusive<char>) -> bool {
        matches!(self.next_char(), range)
    }

    fn next_char(&self) -> Option<char> {
        self.payload.chars().nth(self.index + 1)
    }

    fn current_char(&self) -> Option<char> {
        self.payload.chars().nth(self.index)
    }

    fn forward(&mut self, positions: usize) {
        self.index = self.index + positions;
    }

    fn current_char_expected(&self, val: char) {
        if let Some(c) = self.current_char() {
            if c != val {
                todo!("Require error expected the char {}", val);
            }
        } else {
            todo!("Require error: cannot found current char");
        }
    }

    fn get_by_span(&self, span: &Span) -> String {
        unsafe { self.payload.get_unchecked(span.start..span.end).to_string() }
    }
}

pub struct Tokenizer {}

impl Tokenizer {
    pub fn parse<A>(payload: A) -> Vec<Token>
    where
        A: ToString,
    {
        let ref mut cursor = Cursor::new(0, payload);
        return Self::parse_by_cursor(cursor);
    }

    fn parse_by_cursor(cursor: &mut Cursor) -> Vec<Token> {
        let mut tokens: Vec<Token> = vec![];

        while cursor.has_current() {
            if cursor.current_matches_char('#') {
                tokens.extend(Self::parse_comment(cursor));
                continue;
            }

            if cursor.current_matches_char('\n') {
                tokens.extend(Self::parse_new_line(cursor));
                continue;
            }

            dbg!(&cursor);
            dbg!(&cursor.current_char());
            dbg!(&tokens);
            todo!();
        }

        tokens
    }

    fn parse_comment(cursor: &mut Cursor) -> Vec<Token> {
        cursor.current_char_expected('#');
        let span_start = cursor.index;

        while cursor.has_current() {
            cursor.forward(1);
            if cursor.current_matches_char('\n') {
                break;
            }
        }

        let span = Span {
            start: span_start,
            end: cursor.index,
        };
        let comment_token = Token::new("comment", &cursor, span);
        vec![comment_token]
    }

    fn parse_new_line(cursor: &mut Cursor) -> Vec<Token> {
        cursor.current_char_expected('\n');
        let span_start = cursor.index;
        cursor.forward(1);
        let span = Span {
            start: span_start,
            end: cursor.index,
        };
        let newline_token = Token::new("newline", &cursor, span);
        vec![newline_token]
    }
}
