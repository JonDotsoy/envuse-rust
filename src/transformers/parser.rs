use super::super::errors::parser_error::ParseError;

use super::super::parser::ast::Variable;
use super::transformer_list::TransformerList;
use super::value_types::ValueType;
use std::collections::BTreeMap;

pub struct Parser;

impl Parser {
    pub fn to_parse_variable(
        transformer_list: &TransformerList,
        variable: &Variable,
        envs: &BTreeMap<String, Option<String>>,
    ) -> Result<ValueType, ParseError> {
        let default_type = String::from("String");
        let transform_type = variable
            .variable_type
            .clone()
            .unwrap_or(default_type)
            .to_lowercase();

        let transformer = match transformer_list.get(&transform_type) {
            Some(transformer) => transformer,
            _ => {
                do yeet ParseError::new(
                    format!("Type {} is not valid type", transform_type),
                    variable.span,
                )
            }
        };

        let value_env = envs.get(&variable.name).unwrap_or(&None);

        if variable.nullable && value_env.is_none() && variable.default_value.is_none() {
            return Ok(ValueType::Null);
        }

        let value_to_transform = if let Some(value_env) = value_env {
            value_env.to_string()
        } else if let Some(expression) = variable.default_value.as_ref() {
            if let Some(default_value) = expression.as_default_value() {
                default_value.value.to_string()
            } else {
                do yeet ParseError::new(format!("Expression cannot found error"), variable.span)
            }
        } else {
            do yeet ParseError::new(
                format!("{} value cannot be null", &variable.name),
                variable.span,
            )
        };

        Ok(transformer.as_ref().parse(value_to_transform))
    }
}
