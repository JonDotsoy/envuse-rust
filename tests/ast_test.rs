#[cfg(test)]
mod ast_test {
    use envuse_parser::parser::{ast::AST};
    use envuse_parser::parser::tokenizer::Tokenizer;
    use insta::assert_debug_snapshot;

    #[test]
    fn parse_executable() {
        let payload = "#!/bin/sh";
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_comment() {
        let payload = "# I'm comment";
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_comment_block() {
        let payload = "# I'm comment\n# Second line comment";
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_comment_multiple_block() {
        let payload = "# I'm comment\n\n# Second line comment";
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_comment_multiple_lines() {
        let payload = "# I'm comment\n    # Second line comment";
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_comment_multiple_block_2() {
        let payload = "# I'm comment\n\n     # Second line comment";
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_variables_name() {
        let payload = "
            foo
        ";
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_variables_with_comment() {
        let payload = "
            # comment
            foo
        ";
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_variables_with_type() {
        let payload = "
            foo: string
        ";
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_variable_with_default_value() {
        let payload = r#"foo = "abc""#;
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_variable_with_default_value_number() {
        let payload = r#"port: number = 3000"#;
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_variable() {
        let payload = r#"
            # comment
            foo: string = "abc"
        "#;
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }

    #[test]
    fn parse_sample_document() {
        let payload = r#"
            ## Document comment
            ## second line

            # comment
            foo: string = "abc"
            bax: number = "123"
            port: number = 3_000
        "#;
        let tokens = Tokenizer::parse(payload);
        assert_debug_snapshot!(AST::parse(tokens));
    }
}
