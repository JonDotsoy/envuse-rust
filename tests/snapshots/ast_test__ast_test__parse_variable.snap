---
source: tests/ast_test.rs
expression: "AST::parse(tokens)"
---
Document {
    span: Span {
        start: 0,
        end: 46,
    },
    executable: None,
    elements: [
        Variable {
            span: Span {
                start: 35,
                end: 46,
            },
            comment: Some(
                CommentBlock {
                    span: Span {
                        start: 13,
                        end: 22,
                    },
                    raw: [
                        "# comment",
                    ],
                },
            ),
            name: "foo",
            variable_type: Some(
                "string",
            ),
            default_value: Some(
                DefaultValue {
                    span: Span {
                        start: 50,
                        end: 53,
                    },
                    value: "abc",
                },
            ),
        },
    ],
}
